lib/spacesuit/cors_middleware.ex

```diff
@@ -0,0 +1,171 @@
+defmodule Spacesuit.CorsMiddleware do
+  require Logger
+  use Elixometer
+
+  @http_server                    Application.get_env(:spacesuit, :http_server)
+  @cors                           Application.get_env(:spacesuit, :cors)
+  @supported_http_methods         [:GET, :POST, :PUT, :PATCH, :DELETE, :HEAD, :OPTIONS]
+  @access_control_request_headers Enum.map(@cors[:access_control_request_headers] || [], fn(s) -> String.downcase(s) end)
+                                  |> Enum.into(MapSet.new)
+
+  @timed(key: "timed.corsMiddleware-execute", units: :millisecond)
+  def execute(req, env) do
+    result = with :ok          <- verify_enabled(),
+         :ok                   <- validate_path(req),
+         {:ok, origin, method} <- parse_cors_request(req),
+         {:ok, headers}        <- get_cors_headers(origin, method, req) do
+      {:ok, @http_server.set_resp_headers(headers, req), env}
+    else
+      {:ok, headers, :OPTIONS} ->
+        with_resp_headers = @http_server.set_resp_headers(headers, req)
+        @http_server.reply(200, with_resp_headers, req)
+        {:stop, with_resp_headers}
+      :skip ->
+        Logger.debug "no CORS headers set for #{req[:method]} #{req[:path]}"
+        {:ok, req, env}
+      {:error, :invalid_cors_request} ->
+        headers = req[:headers]
+        Logger.warn "Invalid CORS request;Origin=#{headers["origin"]};Method=#{req[:method]};ACCESS_CONTROL_REQUEST_HEADERS=#{headers["access_control_request_headers"]}"
+        msg = Spacesuit.ApiMessage.encode(
+          %Spacesuit.ApiMessage{status: "error", message: "Invalid CORS request"}
+        )
+        @http_server.reply(403, %{}, msg, req)
+        {:stop, req}
+    end
+    result
```

Comment:
I'm having trouble understanding what is supposed to come out of this in some cases. The logic in `get_cors_headers` is very complicated and the output is fed back into this `with`.